frontend-react/src/hooks/filters/UseCursorManager.ts

```diff
@@ -1,110 +1,95 @@
-    }, [cursors, currentIndex]);
-    const [hasPrev, setHasPrev] = useState<boolean>(false);
-    const [hasNext, setHasNext] = useState<boolean>(false);
-
-    /* Handles verifying cursor duplication */
-    const cursorExists = useCallback(
-        (c: string) => {
-            return Array.from(cursors.values()).includes(c);
-        },
-        [cursors]
-    );
-
-    /* Checks for next and previous values in map from currentIndex OR from
-     * passed-in index. */
-    const checkNextPrev = useCallback(
-        (i?: number) => {
-            setHasNext(cursors.get((i || currentIndex) + 1) !== undefined);
-            setHasPrev(cursors.get((i || currentIndex) - 1) !== undefined);
-        },
-        [currentIndex, cursors]
-    );
-
-    /* Private method for handling cursor addition since we may not always
-     * want to just append a cursor. Checks for duplicates and won't add
-     * duplicate cursors. */
-    const addCursorNoDuplicates = (index: number, val: string) => {
-        if (!cursorExists(val)) {
-            setCursors(cursors.set(index, val));
-        }
-    };
-
-    /* Handles simple appending of new cursors to the end of the Map */
-    const addNextCursor = (val: string) => {
-        addCursorNoDuplicates(cursors.size, val);
-        checkNextPrev();
-    };
+type CursorReducer<Cursors, CursorAction> = (
+    cursors: Cursors,
+    action: CursorAction
+) => Cursors;
 
-    /* Handles navigating the cursor map */
-    const goTo = (i: number) => {
-        if (i >= cursors.size) {
-            setCurrentIndex(currentIndex);
-        } else {
-            setCurrentIndex(i);
-            checkNextPrev(i);
-        }
-    };
+interface CursorManager {
+    cursors: Cursors;
+    hasPrev: boolean;
+    hasNext: boolean;
+    update: React.Dispatch<CursorAction>;
+}
 
-    /* Can reset the cursor map and current index  */
-    const resetCursors = (newFirst?: string) => {
-        setCursors(
-            new Map<number, string>([[0, newFirst || firstCursor || ""]])
-        );
-        setCurrentIndex(0);
-    };
+const cursorReducer = (state: Cursors, action: CursorAction): Cursors => {
+    const { type, payload } = action;
+    switch (type) {
+        case CursorActionType.PAGE_UP:
+            /* Pages up by setting new current with state.next,
+             * sets next to an empty string since ADD_NEXT handles
+             * that, and adds state.current to the history. */
+            return {
+                current: state.next,
+                next: "",
+                history: [...state.history, state.current],
+            } as Cursors;
+        case CursorActionType.PAGE_DOWN:
+            /* Pages down by popping the last history item as current,
+             * sets next to state.current, and spreads the remaining
+             * history */
+            return {
+                current: state.history.pop(),
+                next: state.current,
+                history: [...state.history],
```

Comment:
Why isn't this trimming of the last item of history?
If you go PAGE_DOWN, PAGE_UP, PAGE_DOWN, PAGE_UP... does the history just keep getting longer and loger?

Reply:
Setting current to `state.history.pop()` actually does handle the trimming! The test case shows a slew of page changes that result in an eventual empty history array when paged back to the first page :)

Reply:
Ha! I didn't see the pop above. Looks GREAT!